cmp.js

```diff
@@ -0,0 +1,71 @@
+'use strict'
+
+var Buffer = require('safe-buffer').Buffer
+var toString = Object.prototype.toString
+
+function cmp (a, b) {
+  // Same logic as https://www.w3.org/TR/IndexedDB-2/#key-construct
+  var ta = type(a)
+  var tb = type(b)
+
+  if (ta !== tb) {
+    // Sort order: number, date, string (lexicographic), binary (bitwise), array (componentwise).
+    if (ta === 'array' && (tb === 'binary' || tb === 'string' || tb === 'date' || tb === 'number')) return 1
+    if (tb === 'array' && (ta === 'binary' || ta === 'string' || ta === 'date' || ta === 'number')) return -1
+    if (ta === 'binary' && (tb === 'string' || tb === 'date' || tb === 'number')) return 1
+    if (tb === 'binary' && (ta === 'string' || ta === 'date' || ta === 'number')) return -1
+    if (ta === 'string' && (tb === 'date' || tb === 'number')) return 1
+    if (tb === 'string' && (ta === 'date' || ta === 'number')) return -1
+    if (ta === 'date' && (tb === 'number')) return 1
+
+    /* istanbul ignore else (because behavior of unsupported types is undefined atm) */
+    if (tb === 'date' && (ta === 'number')) return -1
+  } else {
+    if (ta === 'binary') return bitwise(a, b)
+    if (ta === 'array') return componentwise(a, b)
+  }
+
+  return a < b ? -1 : a > b ? 1 : 0
+}
+
+module.exports = cmp
+
+function bitwise (a, b) {
```

Comment:
maybe should use [Buffer.compare](https://www.w3schools.com/nodejs/met_buffer_compare.asp) as this has the potential to be hardware accelerated? I guess safe-buffer could be changed to implement compare with this bitwise code if its missing?

Reply:
I wonder if this could be sped up with an Uint32Array (as a zero-copy view) too?

Reply:
A limitation of `Buffer.compare` is that it can only compare Buffers (and Uint8Arrays depending on version; browserified `buffer` has that since v5). It could be faster, worth a benchmark. For small buffers it probably won't matter, might even hurt performance.

> I wonder if this could be sped up with an Uint32Array (as a zero-copy view) too?

How so?

Reply:
```
function cmp(a, b){
  let minLen = Math.min(a.byteLength, b.byteLength);

  if(minLen % 4 == 0 && a.byteOffset % 4 == 0 && b.byteOffset % 4 == 0){
    let _a = new Uint32Array(a.buffer, a.byteOffset, minLen);
    let _b = new Uint32Array(a.buffer, a.byteOffset, minLen);

    for(let i=0; i< minLen; i++){
      var c = _a[i] - _b[i]
      if (c) return c
    }
    return a.length - b.length
  }
  else if(minLen % 2 == 0 && a.byteOffset % 2 == 0 && b.byteOffset % 2 == 0){
    let _a = new Uint16Array(a.buffer, a.byteOffset, minLen);
    let _b = new Uint16Array(a.buffer, a.byteOffset, minLen);

    for(let i=0; i< minLen; i++){
      var c = _a[i] - _b[i]
      if (c) return c
    }
    return a.length - b.length
  }
  else {
    return a.compare(b);
  }
}
```

... for example; should be the same result but far fewer iterations when you're dealing with large keys. I imagine you'd only want to do this if the keys were past a certain size - to offset the need to do modulo 3 times.

Reply:
dam no it won't work; I think it requires BigEndian

Reply:
Could you explain your theory behind it? I don't quite follow

Reply:
ahh ... it does work to do the above. but `buffer.compare` doesn't have the same output when the lengths are different.

Reply:
the idea is that rather than compare loop over and access 80 elements, you can can (if the buffers are aligned properly) be wrapped with a Uint32Array and you reduce the access and loops to 20. however, its dependent on the Endianness of the system to ensure a < b

Reply:
more basic example: 
```
let a = Buffer.from([0, 1, 2, 4]).readUInt32BE(0);
let b = Buffer.from([1, 1, 2, 4]).readUInt32BE(0);

console.log(a, b, "a < b", a < b);
```

Reply:
https://gist.github.com/vweevers/dbcb1c88dc9e1f78b238df1026aa86ed

Benchmark comparing two sequential keys:

```
node v10.14.1
simple loop x 4,099,738 ops/sec ±0.23% (91 runs sampled)
Buffer.compare x 12,635,054 ops/sec ±0.30% (94 runs sampled)
Uint32Array x 4,413,811 ops/sec ±0.31% (97 runs sampled)
Fastest is Buffer.compare
```

Benchmark comparing two random keys (no surprise here):

```
node v10.14.1
simple loop x 174,031,665 ops/sec ±0.30% (94 runs sampled)
Buffer.compare x 15,825,608 ops/sec ±0.45% (95 runs sampled)
Uint32Array x 9,217,284 ops/sec ±0.36% (94 runs sampled)
Fastest is simple loop
```

Reply:
And on my machine, keys need to be at least 64 bytes long for `Buffer.compare` to become faster.

Reply:
> but I'm not doing benchmarks yet

ah dammit

Reply:
sequential for me with a fix (` for (let i = 0; i < _a.length; i++) ... `)  on `cmpUint32Array': 
```shell
node v10.16.0
simple loop x 2,268,015 ops/sec ±4.10% (82 runs sampled)
Buffer.compare x 11,689,346 ops/sec ±0.71% (94 runs sampled)
Uint32Array x 3,400,119 ops/sec ±0.65% (92 runs sampled)
Fastest is Buffer.compare
```

Reply:
So a definite no with Uint32Array. and a _maybe_ on buffer.compare. :D thanks for testing.